packages/devtools_app/lib/src/screens/memory/panes/leaks/controller.dart

```diff
@@ -6,78 +6,81 @@ import 'dart:async';
 import 'package:vm_service/vm_service.dart';
 
 import '../../../../config_specific/import_export/import_export.dart';
-import '../../../../config_specific/logger/logger.dart' as logger;
 import '../../../../primitives/utils.dart';
-import '../../../../service/service_extensions.dart';
 import '../../../../shared/globals.dart';
 import '../../primitives/memory_utils.dart';
 import 'diagnostics/formatter.dart';
 import 'diagnostics/leak_analyzer.dart';
 import 'diagnostics/model.dart';
-import 'instrumentation/model.dart';
 import 'primitives/analysis_status.dart';
-
-// TODO(polina-c): reference these constants in dart SDK, when it gets submitted
-// there.
-// https://github.com/flutter/devtools/issues/3951
-const _extensionKindToReceiveLeaksSummary = 'memory_leaks_summary';
-const _extensionKindToReceiveLeaksDetails = 'memory_leaks_details';
+import 'primitives/simple_items.dart';
 
 const yamlFilePrefix = 'memory_leaks';
 
 class LeaksPaneController {
-  LeaksPaneController() {
-    _subscribeForMemoryLeaksMessages();
+  LeaksPaneController()
+      : assert(
+          supportedLeakTrackingProtocols
+              .contains(appLeakTrackerProtocolVersion),
+        ) {
+    subscriptionWithHistory = serviceManager
+        .service!.onExtensionEventWithHistory
+        .listen(_onAppMessageWithHistory);
   }
 
-  final status = AnalysisStatusController();
+  final analysisStatus = AnalysisStatusController();
 
   final leakSummaryHistory = ValueNotifier<String>('');
-  final leakSummaryReceived = ValueNotifier<bool>(false);
+  late String appProtocolVersion;
+  final appStatus =
+      ValueNotifier<AppStatus>(AppStatus.noCommunicationsRecieved);
+
   LeakSummary? _lastLeakSummary;
 
   final _exportController = ExportController();
 
-  late StreamSubscription summarySubscription;
-  late StreamSubscription detailsSubscription;
-
-  /// Subscribes for summary with history and for details without history.
-  void _subscribeForMemoryLeaksMessages() {
-    detailsSubscription =
-        serviceManager.service!.onExtensionEvent.listen(_receivedLeaksDetails);
-
-    summarySubscription = serviceManager.service!.onExtensionEventWithHistory
-        .listen(_receivedLeaksSummary);
-  }
+  late StreamSubscription subscriptionWithHistory;
 
   void dispose() {
-    unawaited(summarySubscription.cancel());
-    unawaited(detailsSubscription.cancel());
-    status.dispose();
+    unawaited(subscriptionWithHistory.cancel());
+    analysisStatus.dispose();
   }
 
-  void _receivedLeaksSummary(Event event) {
-    if (event.extensionKind != _extensionKindToReceiveLeaksSummary) return;
-    leakSummaryReceived.value = true;
-    try {
-      final newSummary = LeakSummary.fromJson(event.json!['extensionData']!);
-      final time = event.timestamp != null
-          ? DateTime.fromMicrosecondsSinceEpoch(event.timestamp!)
-          : DateTime.now();
+  static DateTime _fromEpochOrNull(int? microsecondsSinceEpoch) =>
```

Comment:
Nit: maybe `_fromEpochOrNow`? `_fromEpochOrNull` is a similar name to `firstWhereOrNull` which returns null, whereas `_fromEpochOrNull` doesn't have a nullable return type.

Reply:
deleted this method